xla/service/gpu/transforms/gemm_rewriter.cc

```diff
@@ -361,28 +362,67 @@ std::optional<MatchedFp8Param> MatchFp8Param(HloInstruction *instr) {
 // dimension. There must be only one contracting and only one non-contracting
 // dimension. Keeps the layout the same.
 HloInstruction *TransposeMatrix(HloInstruction *instr, int64_t contracting_dim,
-                                absl::Span<const int64_t> batch_dims) {
+                                absl::Span<const int64_t> batch_dims,
+                                bool col_maj = false) {
   // Identify the dimensional order which describes a transpose of the
   // contracting and non-contracting dimensions of the GEMM.
   std::vector<int64_t> permutation(instr->shape().dimensions_size(), -1);
-  // Discard the batch dimensions.
-  for (int64_t batch_dim : batch_dims) {
-    permutation[batch_dim] = batch_dim;
-  }
   // Identify the non-contracting dimension.
   int non_contracting_dim;
   for (int i = 0; i < instr->shape().dimensions_size(); ++i) {
     if (permutation[i] == -1 && contracting_dim != i) {
       non_contracting_dim = i;
     }
   }
-  permutation[non_contracting_dim] = contracting_dim;
-  permutation[contracting_dim] = non_contracting_dim;
+  if (!col_maj) {
+    // Discard the batch dimensions.
+    for (int64_t batch_dim : batch_dims) {
+      permutation[batch_dim] = batch_dim;
+    }
+
+    permutation[non_contracting_dim] = contracting_dim;
+    permutation[contracting_dim] = non_contracting_dim;
+
+    Shape new_shape = ShapeUtil::PermuteDimensions(permutation, instr->shape());
+    *new_shape.mutable_layout() = instr->shape().layout();
+
+    return instr->AddInstruction(
+        HloInstruction::CreateTranspose(new_shape, instr, permutation));
+  }
+
+  Shape normalized_input_shape =
```

Comment:
By going to the normalized input shape, we essentially apply the permutation InversePermutation(Reverse(instr->shape().layout().minor_to_major())) to the shape.

I see that the loop below computes new_contracting_dim and new_non_contracting_dim correctly, but it does not apply any permutation to the batch dimensions. Is this ok, as we don't care about the order of the batch dimensions? Or is it a bug?

Reply:
Yes, it is intended that the batch dims shouldn't be permuted. E.g. a is the input tensor with non-normalized shape. where batch_dims = [1,2], non_contracting_dim=[0] and contracting_dim=[3] , d is output.
```
a = [2,3,4,5]{0,2,3,1}
a0 = [4,3,5,2]{3,2,1,0}bitcast(a)
c = [4,3,2,5]{3,2,1,0}transpose(a0),dimensions=(0,1,3,2)
d =[5,3,4,2]{0,2,3,1}bitcast(c)
```
since the transpose is applied on a0 which has descending layout already, the it's ok to leave batch dims alone because they are also in descending order.